File: tools/zz_chrome_dump.c
/* tools/zz_chrome_dump.c — chrome.c 純粋部の差分 fuzz 用 C オラクル dump ドライバ。
 *
 * 役割: stdin の命令行を 1 行ずつ実行し、結果を機械可読な 1 行で stdout に返す。
 * Rust 側 `rust/ifuto-core/examples/zz_chrome_dump.rs` と規格を合わせ、
 * tools/zz_chrome_diff.py が両出力を byte 突合する。
 *
 * 規約:
 *   - 文字列引数は lowercase hex
 *   - 数値は 10 進（負数可）
 *   - fs.exists は決定論予言者擬装: exists(s) = (fnv1a64(s) % 4 == 0)
 *     （Rust 側が同一式で実装）
 *   - DUP の cap == 0 は C が u32 underflow で 4GiB alloc に進む領域のため生成禁止
 *     （chrome.rs の dup_cap 契約と同じ信頼域）
 *
 * 命令一覧（→ 出力）:
 *   CI  <hay> <needle>              → CI <0|1>
 *   DUP <s> <cap>                   → DUP <out-hex>
 *   SCR <scroll> <delta> <vh> <doc_h> → SCR <new-scroll>
 *   SCT <pos> <vh> <doc_h>          → SCT <new-scroll>
 *   QUI <n_tabs> <armed_at> <now>   → QUI <rc> <new-armed_at>
 *   LNK <idx> <delta> <n_links>     → LNK <new-idx>
 *   RES <input> <cwd> <cap>         → RES <rc> <out-hex|->
 *   FT  <max> <ntab> <query> [t u g]×ntab → FT <i0,i1,...>   (g が "-" なら NULL)
 *
 * トークンは空白（スペース/タブ/改行）区切り。巨大 hex トークン（数 KB 級）も
 * sscanf による中間コピーなしで span から直接 unhex する（截断事故の排除）。
 */
#include "../src/chrome.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* ---- 決定論 fs 予言者（Rust 側と同一式にすること） ---- */
static u64 fnv1a64(const char *s) {
    u64 h = 14695981039346656037ULL;
    while (*s) { h ^= (unsigned char)*s++; h *= 1099511628211ULL; }
    return h;
}
static bool fake_exists(const char *path, void *ctx) {
    (void)ctx;
    return fnv1a64(path) % 4 == 0;
}
static IfStr fake_read(IfArena *a, const char *path, void *ctx) {
    (void)a; (void)path; (void)ctx;
    return if_str(NULL, 0);
}
static bool fake_write(const char *path, const void *buf, size_t n, void *ctx) {
    (void)path; (void)buf; (void)n; (void)ctx;
    return true;
}
static bool fake_mkpath(const char *dir, void *ctx) {
    (void)dir; (void)ctx;
    return true;
}

/* ---- トークナイザ（空白区切り、NUL 終端 span を返す） ---- */
static char *next_tok(char **pp) {
    char *p = *pp;
    while (*p == ' ' || *p == '\t' || *p == '\n' || *p == '\r') p++;
    if (!*p) { *pp = p; return NULL; }
    char *start = p;
    while (*p && *p != ' ' && *p != '\t' && *p != '\n' && *p != '\r') p++;
    if (*p) *p++ = 0;   /* 区切りを NUL に置換して次へ */
    *pp = p;
    return start;
}
static long tok_long(char **pp) {
    char *t = next_tok(pp);
    return t ? strtol(t, NULL, 10) : 0;
}

/* ---- hex ユーティリティ ---- */
static int hexval(int c) {
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    return -1;
}
/* hex トークンをバッファへ（常に NUL 終端。戻り値はバイト数） */
static size_t unhex(const char *s, char *out, size_t cap) {
    size_t n = 0;
    while (s[0] && s[1] && n + 1 < cap) {
        int hi = hexval((unsigned char)s[0]);
        int lo = hexval((unsigned char)s[1]);
        if (hi < 0 || lo < 0) break;
        out[n++] = (char)((hi << 4) | lo);
        s += 2;
    }
    out[n] = 0;
    return n;
}
/* トークンから直接 unhex（見つからなければ空） */
static size_t unhex_tok(char **pp, char *out, size_t cap) {
    char *t = next_tok(pp);
    if (!t) { out[0] = 0; return 0; }
    return unhex(t, out, cap);
}
static void put_hex(const char *s, size_t n) {
    static const char *D = "0123456789abcdef";
    for (size_t i = 0; i < n; i++)
        printf("%c%c", D[(unsigned char)s[i] >> 4], D[(unsigned char)s[i] & 15]);
}

#define BUFSZ 40960
static char IBUF[BUFSZ], OBUF[BUFSZ], ABUF[BUFSZ], BBUF[BUFSZ];

int main(void) {
    static char line[BUFSZ * 3];
    IfFsOps fs = { 0 };
    fs.exists = fake_exists;
    fs.read_file = fake_read;
    fs.write_file = fake_write;
    fs.append = fake_write;
    fs.mkpath = fake_mkpath;

    static char raw[BUFSZ * 3];
    while (fgets(line, sizeof line, stdin)) {
        memcpy(raw, line, sizeof raw);
        char *p = line;
        char *op = next_tok(&p);
        if (!op) continue;
        if (!strcmp(op, "CI")) {
            unhex_tok(&p, ABUF, sizeof ABUF);
            unhex_tok(&p, BBUF, sizeof BBUF);
            printf("CI %d\n", if_chrome_test_ci_contains(ABUF, BBUF) ? 1 : 0);
        } else if (!strcmp(op, "DUP")) {
            unhex_tok(&p, ABUF, sizeof ABUF);
            u32 cap = (u32)tok_long(&p);
            char *r = if_chrome_test_dup_cap(ABUF, cap);
            printf("DUP ");
            put_hex(r, strlen(r));
            printf("\n");
            free(r);
        } else if (!strcmp(op, "SCR")) {
            long scroll = tok_long(&p), delta = tok_long(&p);
            long vh = tok_long(&p), doc_h = tok_long(&p);
            IfChrome c;
            memset(&c, 0, sizeof c);
            IfTab t;
            memset(&t, 0, sizeof t);
            IfLayout lay;
            memset(&lay, 0, sizeof lay);
            t.lay = &lay;
            t.doc_h = (i32)doc_h;
            t.scroll = (i32)scroll;
            c.tabs[0] = &t;
            c.n_tabs = 1;
            c.active = 0;
            printf("SCR %d\n", if_chrome_scroll(&c, (i32)delta, (i32)vh));
        } else if (!strcmp(op, "SCT")) {
            long pos = tok_long(&p), vh = tok_long(&p), doc_h = tok_long(&p);
            IfChrome c;
            memset(&c, 0, sizeof c);
            IfTab t;
            memset(&t, 0, sizeof t);
            IfLayout lay;
            memset(&lay, 0, sizeof lay);
            t.lay = &lay;
            t.doc_h = (i32)doc_h;
            c.tabs[0] = &t;
            c.n_tabs = 1;
            c.active = 0;
            if_chrome_scroll_to(&c, (i32)pos, (i32)vh);
            printf("SCT %d\n", t.scroll);
        } else if (!strcmp(op, "QUI")) {
            long nt = tok_long(&p), armed = tok_long(&p), now = tok_long(&p);
            IfChrome c;
            memset(&c, 0, sizeof c);
            c.n_tabs = (i32)nt;
            c.quit_armed_at = armed;
            bool r = if_chrome_quit(&c, now);
            printf("QUI %d %lld\n", r ? 1 : 0, (long long)c.quit_armed_at);
        } else if (!strcmp(op, "LNK")) {
            long idx = tok_long(&p), delta = tok_long(&p), nl = tok_long(&p);
            IfChrome c;
            memset(&c, 0, sizeof c);
            IfTab t;
            memset(&t, 0, sizeof t);
            IfLayout lay;
            memset(&lay, 0, sizeof lay);
            lay.n_links = (u32)nl;
            t.lay = &lay;
            t.link_idx = (i32)idx;
            c.tabs[0] = &t;
            c.n_tabs = 1;
            c.active = 0;
            printf("LNK %d\n", if_chrome_link_move(&c, (i32)delta));
        } else if (!strcmp(op, "RES")) {
            unhex_tok(&p, IBUF, sizeof IBUF);
            unhex_tok(&p, ABUF, sizeof ABUF);
            u32 cap = (u32)tok_long(&p);
            IfChrome c;
            memset(&c, 0, sizeof c);
            c.fs = fs;
            /* out バッファを 0x01 センチネルで初期化（C が書かなければ検出できる。
             * fuzz 入力は 0x01 を含まないドメインに制限してある） */
            memset(OBUF, 0x01, 512);
            OBUF[511] = 0;
            i32 rc = if_chrome_resolve(&c, IBUF, ABUF, OBUF, cap);
            printf("RES %d ", rc);
            if ((unsigned char)OBUF[0] == 0x01) {
                printf("-\n");
            } else {
                put_hex(OBUF, strlen(OBUF));
                printf("\n");
            }
        } else if (!strcmp(op, "FT")) {
            long max = tok_long(&p), ntab = tok_long(&p);
            char qhex[8192];
            {
                char *q = next_tok(&p);
                if (!q) goto bad;
                snprintf(qhex, sizeof qhex, "%s", q);
            }
            if (ntab < 0 || ntab > 64) goto bad;
            IfChrome c;
            memset(&c, 0, sizeof c);
            static IfTab tabs[64];
            static char ttitle[64][IF_TITLE_CAP], turl[64][IF_URL_CAP], tgrp[64][IF_GROUP_CAP];
            char gbuf[IF_GROUP_CAP * 2 + 2];
            int i;
            for (i = 0; i < ntab; i++) {
                unhex_tok(&p, ttitle[i], IF_TITLE_CAP);
                unhex_tok(&p, turl[i], IF_URL_CAP);
                char *gt = next_tok(&p);
                if (!gt) goto bad;
                snprintf(gbuf, sizeof gbuf, "%s", gt);
                memset(&tabs[i], 0, sizeof tabs[i]);
                if (!strcmp(gbuf, "-")) {
                    tabs[i].group = NULL;
                } else {
                    unhex(gbuf, tgrp[i], IF_GROUP_CAP);
                    tabs[i].group = tgrp[i];
                }
                tabs[i].title = ttitle[i];
                tabs[i].url = turl[i];
                c.tabs[c.n_tabs++] = &tabs[i];
            }
            if (i != ntab) goto bad;
            unhex(qhex, IBUF, sizeof IBUF);
            i32 idx[64];
            i32 got = if_chrome_find_tabs(&c, IBUF, idx, (i32)max);
            printf("FT ");
            for (i32 k = 0; k < got; k++) printf("%s%d", k ? "," : "", idx[k]);
            printf("\n");
        } else {
        bad:
            printf("BAD %s", raw);
        }
        fflush(stdout);
    }
    return 0;
}

File: rust/ifuto-core/examples/zz_chrome_dump.rs
//! chrome モデル純粋部の Rust 側ダンプドライバ（差分 fuzz 用観測点）。
//!
//! `tools/zz_chrome_dump.c`（C オラクル）と行フォーマットを byte 完全一致させ、
//! `tools/zz_chrome_diff.py` が全行突合する。規約は C 側コメントを参照。
//! （数値は 10 進、文字列は hex。fs.exists は FNV-1a 64bit %4==0 の決定論予言者）
//! `TabSearch` は byte 列保持のため非 UTF-8 入力もそのまま流せる。

use ifuto_core::chrome;
use std::io::{self, BufRead, Write};

/// unhex（C 側と同一規約: 2 文字ずつ、無効文字/奇数端数で打ち切り）
fn unhex(s: &str, out: &mut Vec<u8>) {
    let b = s.as_bytes();
    let mut i = 0;
    while i + 1 < b.len() {
        let hi = (b[i] as char).to_digit(16);
        let lo = (b[i + 1] as char).to_digit(16);
        match (hi, lo) {
            (Some(h), Some(l)) => out.push(((h << 4) | l) as u8),
            _ => break,
        }
        i += 2;
    }
}

fn tohex(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

/// C 側 oracle_exists と同一式（FNV-1a 64bit の下位 2bit が 0 のとき存在）
fn oracle_exists(path: &[u8]) -> bool {
    let mut h: u64 = 0xcbf29ce484222325;
    for &p in path {
        h ^= p as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    (h & 3) == 0
}

fn pi64(t: Option<&str>) -> i64 {
    t.and_then(|s| s.parse::<i64>().ok()).unwrap_or(0)
}

fn main() {
    let stdin = io::stdin();
    let mut out = io::BufWriter::new(io::stdout());

    for line in stdin.lock().lines() {
        let line = line.unwrap();
        let mut it = line.split_whitespace();
        let op = match it.next() {
            Some(o) => o,
            None => continue,
        };
        match op {
            "CI" => {
                let mut h = Vec::new();
                let mut n = Vec::new();
                unhex(it.next().unwrap_or(""), &mut h);
                unhex(it.next().unwrap_or(""), &mut n);
                writeln!(out, "CI {}", chrome::ci_contains(&h, &n) as i32).unwrap();
            }
            // FT <max> <ntab> <query-hex> [t u g]×ntab（g が "-" なら NULL）
            "FT" => {
                let max = pi64(it.next());
                let ntab = pi64(it.next());
                if !(0..=64).contains(&ntab) {
                    // C ドライバと同一の域外拒否
                    writeln!(out, "BAD {line}").unwrap();
                    continue;
                }
                let mut q = Vec::new();
                unhex(it.next().unwrap_or(""), &mut q);
                let mut tabs = Vec::new();
                let mut bad = false;
                for _ in 0..ntab.max(0) {
                    let (mut t, mut u) = (Vec::new(), Vec::new());
                    let g = match (it.next(), it.next(), it.next()) {
                        (Some(th), Some(uh), Some(gh)) => {
                            unhex(th, &mut t);
                            unhex(uh, &mut u);
                            gh
                        }
                        _ => {
                            bad = true;
                            break;
                        }
                    };
                    let group = if g == "-" {
                        None
                    } else {
                        let mut v = Vec::new();
                        unhex(g, &mut v);
                        Some(v)
                    };
                    tabs.push(chrome::TabSearch {
                        title: t,
                        url: u,
                        group,
                    });
                }
                if bad {
                    writeln!(out, "BAD {line}").unwrap();
                    continue;
                }
                let idx = if max <= 0 {
                    Vec::new()
                } else {
                    chrome::find_tabs(&tabs, &q, max as usize)
                };
                let list: Vec<String> = idx.iter().map(|i| i.to_string()).collect();
                // C ドライバと同一書式（"FT " + カンマ区切り。0 件は "FT " のみ）
                writeln!(out, "FT {}", list.join(",")).unwrap();
            }
            "DUP" => {
                let mut s = Vec::new();
                unhex(it.next().unwrap_or(""), &mut s);
                let cap = it.next().unwrap_or("0").parse::<u64>().unwrap_or(0) as u32;
                let r = chrome::dup_cap(&s, cap);
                writeln!(out, "DUP {}", tohex(&r)).unwrap();
            }
            "SCR" => {
                let scroll = pi64(it.next()) as i32;
                let delta = pi64(it.next()) as i32;
                let vh = pi64(it.next()) as i32;
                let doc_h = pi64(it.next()) as i32;
                writeln!(
                    out,
                    "SCR {}",
                    chrome::scroll_apply(scroll, delta, vh, doc_h)
                )
                .unwrap();
            }
            "SCT" => {
                let pos = pi64(it.next()) as i32;
                let vh = pi64(it.next()) as i32;
                let doc_h = pi64(it.next()) as i32;
                writeln!(out, "SCT {}", chrome::scroll_to_apply(pos, vh, doc_h)).unwrap();
            }
            "QUI" => {
                let nt = pi64(it.next()) as i32;
                let armed = pi64(it.next());
                let now = pi64(it.next());
                let (rc, new_armed) = chrome::quit_decide(nt, armed, now);
                writeln!(out, "QUI {} {}", rc as i32, new_armed).unwrap();
            }
            "LNK" => {
                let idx = pi64(it.next()) as i32;
                let delta = pi64(it.next()) as i32;
                let nl = pi64(it.next()) as u32;
                writeln!(out, "LNK {}", chrome::link_move(idx, delta, nl)).unwrap();
            }
            "RES" => {
                let mut input = Vec::new();
                let mut cwd = Vec::new();
                unhex(it.next().unwrap_or(""), &mut input);
                unhex(it.next().unwrap_or(""), &mut cwd);
                let cap = it.next().unwrap_or("0").parse::<u64>().unwrap_or(0) as u32;
                let (rc, written) = chrome::resolve(&input, &cwd, cap, &mut oracle_exists);
                match written {
                    None => writeln!(out, "RES {rc} -").unwrap(),
                    Some(v) => writeln!(out, "RES {} {}", rc, tohex(&v)).unwrap(),
                }
            }
            _ => writeln!(out, "BAD {line}").unwrap(),
        }
        out.flush().unwrap();
    }
}
